runtime/depends_on.cc

```diff
@@ -41,75 +44,191 @@ class dependencies : public recursive_node_visitor {
     }
   }
 
+  template <typename T>
+  void visit_let(const T* op) {
+    for (const auto& p : op->lets) {
+      p.second.accept(this);
+      shadowed.push_back(p.first);
+    }
+    op->body.accept(this);
+    for (const auto& p : op->lets) {
+      shadowed.pop_back();
+    }
+  }
+
+  template <typename T>
+  void visit_sym_body(const T* op) {
+    if (!op->body.defined()) return;
+    shadowed.push_back(op->sym);
+    op->body.accept(this);
+    shadowed.pop_back();
+  }
+
+  void visit(const loop* op) override {
+    op->bounds.min.accept(this);
+    op->bounds.max.accept(this);
+    if (op->step.defined()) op->step.accept(this);
+
+    std::vector<int> old_ref_count(var_deps.size());
+    for (std::size_t i = 0; i < var_deps.size(); ++i) {
+      old_ref_count[i] = var_deps[i].second.ref_count;
+    }
+
+    visit_sym_body(op);
+
+    for (std::size_t i = 0; i < var_deps.size(); ++i) {
+      if (var_deps[i].second.ref_count > old_ref_count[i]) {
+        var_deps[i].second.used_in_loop = true;
+      }
+    }
+  }
+
   void visit(const call_stmt* op) override {
     for (symbol_id i : op->inputs) {
-      if (vars_contains(i)) {
-        result.buffer = true;
-        result.buffer_input = true;
-      }
+      update_deps(i, [](depends_on_result& deps) {
```

Comment:
nit: could use a single local lambda for both of these loops

Reply:
They set different things though? It's probably doable, just not sure it's worth avoiding the very minor duplication.

Reply:
Whoops! You're right, I misread it.